### DISCUSSION TITLE
Make initializer inside using scope

### DISCUSSION THREAD

**Discussion by ahdung** [dotnet/csharplang#7751] [Language Ideas]
![微信截图_20231208114504](https://github.com/dotnet/csharplang/assets/12094247/c4e16237-f70a-4a4d-80b7-cb789d1b5446)

Above codes actually is:
```c#
FolderBrowserDialog folderBrowserDialog = new FolderBrowserDialog();
folderBrowserDialog.ShowNewFolderButton = true;
folderBrowserDialog.Description         = Title;
folderBrowserDialog.SelectedPath        = InitialDirectory;
using (FolderBrowserDialog fbd = folderBrowserDialog)
{
    ...
}
```
So when `folderBrowserDialog.ShowNewFolderButton = true;` occurred exception, `folderBrowserDialog` will not be disposed. So for avoid this issue we have to abandon using **object initializer**:
```c#
using var fbd = new FolderBrowserDialog();
fbd.ShowNewFolderButton = true;
fbd.Description         = Title;
fbd.SelectedPath        = InitialDirectory;
```
which is no longer elegant, so please make initializer statement using-friendly.

Ref: https://dev.to/powerz/dangerous-using-declaration-c-80-3h86

**Comment by yaakov-h** [comment]
This belongs at https://github.com/dotnet/roslyn-analyzers/, however the analyzer is technically correct.

`using var foo = new Foo { Bar = x() };` is translated by the compiler into:

```
var foo = new Foo();
foo.Bar = x();
using (foo) {
```

If `foo.set_Bar` or `x()` throw an exception then `foo` does not get disposed.

**Reply by ahdung** [reply]
I agree it is correct at technically, but it does not meet human expectations.

**Reply by CyrusNajmabadi** [reply]
This is the same issue as if your constructor throws.  The Dispose will nto be called either.

Fundamentally, you have to design and use your types as appropriate.  For example, having a disposable type that has assignable properties, which also may throw is just a *realllllly* bad idea on many levels.  

Instead of trying to language around it. I'd write analyzers to warn you about those types so you can fix them to be better behaved.

**Reply by CyrusNajmabadi** [reply]
Or if some of your arguments end up throwing.  for example:

```
using (var v = new DisposeAllMyArgs(GetArg1(), GetArg2(), GetArg3ButThisThrows())
{
}
```

In this case, Arg1 and Arg2 will not be disposed.  It is a prerequisite for these systems that your construction phase be exception free.  Once hte construction phase is complete, then you are safe from exceptions happening after that point.

**Reply by ahdung** [reply]
@CyrusNajmabadi Your example is different,, when constructor occurred exception:
`using var obj = new WillThrowException();`
Or inner calling occurred exception:
`using var obj = new Foo(WillThrowException());`

We accept this, because they all unfinished construction, the `obj` not born yet, nothing need dispose, but:
`using var obj = new Foo { willThrowException };`
remain a `obj`, the `using` has the responsibility to complete its mission.

**Reply by HaloFour** [reply]
A better example might be:

```cs
using var obj = new SomethingOrOther().SomeMethod();
```

If `SomethingOrOther` is disposable, but `SomeMethod()` throws, the resource is not disposed.  The resource to be disposed is `obj`, and it's not under the protection of `using` until it has been assigned, which is after the entire expression (chained methods and initializers included) is complete.

**Reply by CyrusNajmabadi** [reply]
> We accept this, because they all unfinished construction, 

Which is the same as with the property assignments.  Until they finish, the instance is unfinished.

**Reply by ahdung** [reply]
@HaloFour Sophistry, In your case, `obj` is created by `SomeMethod()`, it has nothing to do with anything else, if `SomeMethod` throws, `obj` will not be created, nothing need dispose, `using` is responsible for `obj`.

**Reply by HaloFour** [reply]
@ahdung 

> if `SomeMethod` throws, `obj` will not be created, nothing need dispose

If `SomeMethod()` throws, the reference created by `new SomethingOrOther()` has still created a disposable resource.  You are correct in that it will not be assigned to `obj`, nor will the `using` statement apply to it.  That is identical to initialization.  The constructor has already been called and the resource allocated, but it has not been assigned to the `using` resource until after all initializers have completed.  In both cases, a thrown exception will result in the allocated resource not being disposed.

**Reply by ahdung** [reply]
@CyrusNajmabadi Putting aside the debate over the definition of "finish", the point is `using` have responsibility for `obj`, as long as `obj` created, `using` need make sure to dispose it. The root of this problem is, `initializer` should happen after construction, but it been compiled to "before".

**Reply by ahdung** [reply]
@HaloFour
> but it has not been assigned to the using resource until after all initializers have completed.

That's the problem is, so I asking for `initializers` should after assignment, the compiled code should equal with:
```c#
using var obj = xxx;
obj.XX = xxx;
```
This is in line with user’s intuitive expectations.

**Reply by HaloFour** [reply]
> This is in line with user’s intuitive expectations.

This is not in line with how assignments have always worked in C#.  The construction and initialization is a part of an expression which is independent of the assignment.

**Reply by CyrusNajmabadi** [reply]
> The root of this problem is, initializer should happen after construction, but it been compiled to "before".

I disagree.  Just with parameters, the execution happens prior.  And if they throw, you won't have the completed object to dispose.  

> That's the problem is, so I asking for initializers should after assignment, the compiled code should equal with:

Strong disagree.  That's not how initialization has ever worked.  If we did that, we would now be able to observe things like:

```c#
this.Foo = new Foo
{
    X = ThrowSomething(),
}
```

Now the `this.Foo` field would be assigned, despite the throwing happening.  Failure to initialize shouldn't end up allowing the value to leak out on the callsite.

**Comment by TahirAhmadov** [comment]
How about:
```cs
var folderBrowserDialog = using new FolderBrowserDialog
{
  ShowNewFolderButton = true,
  Description = Title,
  SelectedPath = InitialDirectory,
};
```
Here, `using new` is a new construct which requires an init block to follow the constructor call, and the above is translated to:
```cs
using var temp = new FolderBrowserDialog();
temp.ShowNewFolderButton = true;
temp.Description = Title;
temp.SelectedPath = InitialDirectory;
var fbd = temp;
```

**Reply by ahdung** [reply]
Same issue here, `initializer` cannot be separated from construction, if they can be separated, then it can achieved directly, not need workaround.

**Reply by TahirAhmadov** [reply]
If you mean that you would like C# to be changed so that `using var folderBrowserDialog = new FolderBrowserDialog { ... };` starts behaving differently, I'm pretty sure that will not happen because it's a huge breaking change.

**Reply by ahdung** [reply]
Breaking change is not problem, context switch can handle this, technical feasibility is, I guess.

**Reply by TahirAhmadov** [reply]
It is an insurmountable problem, and if by "context switch" you mean having some kind of compiler directive or project setting which changes how this behaves, that's 99.99% not happening either, because it creates a language dialect and this issue is nowhere near the bar that's required for a language dialect.

**Reply by CyrusNajmabadi** [reply]
> Breaking change is not problem

It's a def problem. Especially as the new semantics violate a design goal of initialization.  If initialization fails, the assignment should not happen. That alone is a serious problem and a big deal.
